src/ucp/core/ucp_worker.c

```diff
@@ -2360,3 +2390,185 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+    ucp_worker_h worker      = req->send.discard_uct_ep.ucp_worker;
+
+    ucp_trace_req(req, "discard_uct_ep flush completion status %s",
+                  ucs_status_string(status));
+
+    /* don't destroy UCT EP from the flush completion callback, schedule
+     * a progress callback on the main thread to destroy UCT EP */
+    uct_worker_progress_register_safe(worker->uct,
+                                      ucp_worker_discard_uct_ep_destroy_progress,
+                                      req, UCS_CALLBACKQ_FLAG_ONESHOT, &cb_id);
+}
+
+static ucs_status_t
+ucp_worker_discard_uct_ep_pending_cb(uct_pending_req_t *self)
+{
+    uct_worker_cb_id_t cb_id = UCS_CALLBACKQ_ID_NULL;
+    ucp_request_t *req       = ucs_container_of(self, ucp_request_t, send.uct);
+    uct_ep_h uct_ep          = req->send.discard_uct_ep.uct_ep;
+    ucp_worker_h worker      = req->send.discard_uct_ep.ucp_worker;
+    ucs_status_t status;
+
+    status = uct_ep_flush(uct_ep, req->send.discard_uct_ep.ep_flush_flags,
+                          &req->send.state.uct_comp);
+    if (status == UCS_ERR_NO_RESOURCE) {
+        status = uct_ep_pending_add(uct_ep, &req->send.uct, 0);
+        ucs_assert((status == UCS_ERR_BUSY) || (status == UCS_OK));
+        if (status == UCS_ERR_BUSY) {
+            uct_worker_progress_register_safe(worker->uct,
+                                              ucp_worker_discard_uct_ep_progress,
+                                              req, UCS_CALLBACKQ_FLAG_ONESHOT,
+                                              &cb_id);
+        }
+    } else if (status != UCS_INPROGRESS) {
+        /* UCS_OK is handled here as well */
+        ucp_worker_discard_uct_ep_flush_comp(&req->send.state.uct_comp, status);
+        return status;
+    }
+
+    /* the request was added to the UCT pending queue or to the UCT worker
+     * progress, need to return UCS_OK in order to remove the callback from
+     * the previous place (either UCT pending queue or UCT worker progress)
+     * to not invoke it several times */
+    return UCS_OK;
+}
+
+static unsigned ucp_worker_discard_uct_ep_progress(void *arg)
+{
+    ucp_request_t *req  = (ucp_request_t*)arg;
+    ucs_status_t status = ucp_worker_discard_uct_ep_pending_cb(&req->send.uct);
+
+    return !UCS_STATUS_IS_ERR(status);
+}
+
+static uct_ep_h
+ucp_worker_discard_wireup_ep(ucp_worker_h worker,
+                             ucp_wireup_ep_t *wireup_ep,
+                             unsigned ep_flush_flags,
+                             uct_pending_purge_callback_t purge_cb,
+                             void *purge_arg)
+{
+    uct_ep_h uct_ep;
+    int is_owner;
+
+    ucs_assert(wireup_ep != NULL);
+    ucs_assert(purge_cb != NULL);
+
+    uct_ep_pending_purge(&wireup_ep->super.super, purge_cb, purge_arg);
+
+    if (wireup_ep->aux_ep != NULL) {
+        /* make sure that there is no WIREUP MSGs anymore */
+        uct_ep_pending_purge(wireup_ep->aux_ep,
```

Comment:
shouldn't we do this before uct_ep_pending_purge()? or not at all?
ucp_wireup_ep_pending_purge already can purge aux_ep